Introduce fluid property abstraction

TimothyEDawson opened the pull request:
Closes #11

A new abstract base class called FluidPhysics was created, from which
all classes which handle fluid property evaluations will derive. A
dataclass called FluidState is introduced to store the thermodynamic
properties within the domain, and is passed around to the property
evaluations to simplify their calls.

This refactoring touches many aspects of the code simultaneously. This commit will:
- Rewrite initialization routines to create a new FluidState object,
  and remove side-effects (i.e. mesh setup).
- Include mass fractions and flamelet scalars in the FluidState, while
  using "composition" to represent the set of transported scalars.
- Introduce CanteraInterface which utilizes Cantera SolutionArray
  objects to handle all evaluations in a vectorized fashion.
- Change ThermoTable to a subclass of CanteraInterface so that any
  evaluations which are not included in the table (e.g. transport
  properties) will automatically fall back to Cantera evaluations.
- Update Combustor class to work with the new physics, and minimize
  conditionals which depend on which physics object is being used.
- Update viscous_flux, plot, probe, and examples to work with the new physics.
- Generalize the hydraulic diameter calculation in the Combustor class to include the approach used by ShockTube.
- Turn ShockTube into a subclass of Combustor to remove duplicated code.

There is still a lot of potential improvements to be made. For example:
- The calling class is responsible for ensuring the properties are
  properly updated and synchronized, and some calculations are repeated.
  This should be made more user-friendly and less error-prone.
- Many convenience functions could be added to FluidState, such as
  slicing to generate a FluidState representing a subset of the domain.
- Ideally the advance_chemistry_FPV and advance_chemistry_FRC would be
  collapsed into a single advance function, with the relevant
  differences subsumed into FluidPhysics calls (e.g. get_source_terms).
- Considering moving the conservative variables into the FluidState as
  well, and rewriting various function calls to accept FluidState
  objects instead of specific variables.

TimothyEDawson committed:
Introduce fluid physics abstraction.

A new abstract base class called FluidPhysics was created, from which
all classes which handle fluid property evaluations will derive. A
dataclass called FluidState is introduced to store the thermodynamic
properties within the domain, and is passed around to the property
evaluations to simplify their calls.

This refactoring touches many aspects of the code simultaneously. This
initial commit will:
- Rewrite initialization routines to create a new FluidState object,
  and remove side-effects (i.e. mesh setup).
- Include mass fractions and flamelet scalars in the FluidState, while
  using "composition" to represent the set of transported scalars.
- Introduce CanteraInterface which utilizes Cantera SolutionArray
  objects to handle all evaluations in a vectorized fashion.
- Change ThermoTable to a subclass of CanteraInterface so that any
  evaluations which are not included in the table (e.g. transport
  properties) will automatically fall back to Cantera evaluations.
- Update Combustor class to work with the new physics, and minimize
  conditionals which depend on which physics object is being used.
- Update viscous_flux to work with the new physics.

There is still a lot of work to do, and potential improvements to be
made. For example:
- The calling class is responsible for ensuring the properties are
  properly updated and synchronized, and some calculations are repeated.
  This should be made more user-friendly and less error-prone.
- Many convenience functions could be added to FluidState, such as
  slicing to generate a FluidState representing a subset of the domain.
- Ideally the advance_chemistry_FPV and advance_chemistry_FRC would be
  collapsed into a single advance function, with the relevant
  differences subsumed into FluidPhysics calls (e.g. get_source_terms).
- Considering moving the conservative variables into the FluidState as
  well, and rewriting various function calls to accept FluidState
  objects instead of specific variables.

TimothyEDawson committed:
Update laminar_flame test case for new physics.

Added option to switch between ThermoTable and CanteraInterface for the
physics. The test case currently does not pass, so sources of
differences need to be identified.

TimothyEDawson committed:
Move ThermoTable back out of the thermo subfolder.

TimothyEDawson committed:
Update test_thermo for new FluidPhysics.

TimothyEDawson committed:
Fix missed FluidPhysics issues in Combustor.

laminar_flame case still does not pass.

TimothyEDawson committed:
Reinitialize mesh parameters.

This fixed a bug where the default dx value was retained after the mesh
was manually set, leading to incorrect calculations of the stable time
step.

TimothyEDawson committed:
Generalize the hydraulic diameter calculation.

Combustor class can now support the hydraulic diameter definition used
by ShockTube class. Also corrected an inconsistency in said definition
where the characteristic length was not being set correctly.

TimothyEDawson committed:
Replace lambda_over_cv with thermal_conductivity.

The lambda_over_cv calls were confusing and used inconsistently
throughout the codebase. Now, only directly querying
thermal_conductivity is supported.

TimothyEDawson committed:
Replace isinstance queries with is_flamelet property.

TimothyEDawson committed:
Update plot and probe functions for FlamePhysics.

TimothyEDawson committed:
Make StanShock a subclass of Combustor.

This is a temporary move to avoid a lot of updates to duplicated code.

TimothyEDawson committed:
Speed up ThermoTable by avoiding set_state.

Calls to set_state include setting up CanteraSolution objects in order
to compute transport properties. This should be avoided unless needed
because it can be expensive. With this commit ThermoTable is quite a bit
faster, but it is still significantly slower than the original version.

TimothyEDawson committed:
Update example cases for new FluidPhysics.

TimothyEDawson committed:
Fix formatting.

TimothyEDawson committed:
Remove unnecessary set_state calls in Combustor.

TimothyEDawson committed:
Return and formalize missing pressure calculation.

TimothyEDawson committed:
Implement new thermo caching strategy.

TimothyEDawson commented:
I believe this is ready for a review. A couple notes:

- Small discrepancies within the Cantera calls lead to results which differ from the reference solution data in test_examples.py.
  - I don't believe it's worth trying to force these to match in general, and instead we should focus on unit tests outside of the example cases.
  - There were two specific root causes of the discrepancies:
    - It turns out that sometimes setting gas.TP and then gas.Y can give slightly different results than setting gas.TPY. ct.SolutionArray objects only use TPY setters.
    - The thermal diffusivity was being computed with cv in the denominator instead of cp, but was multiplied by cp to get the thermal conductivity. This was corrected.
- I did need to force a cache invalidation in a few spots to ensure it takes the same path to computing transport properties as the original solver did. There are better ways to handle this, but I'll open up separate Issues for that work.

MatthewBonanni reviewed (approved):
I really like how this is structured. Looks great!

TimothyEDawson commented:
Spotted a few mistakes in FPVTable, I'll push a fix shortly and then merge this in.

TimothyEDawson committed:
Fix minor issues in FPVTable.

Added a lookup_direct option, mainly for compatibility with JICF model
logic which would be cumbersome to rewrite for the new interface.

Fixed an issue where n_scalars and scalar_names, which were decorated
functions in FluidPhysics, could not be overwritten in FPVTable.

TimothyEDawson committed:
Move fuel/ox/prog definitions into FluidPhysics.

FPVTable can now extract the fuel and oxidizer compositions from the
flamelet table automatically if not provided. Something similar could
likely be done for the progress variable.

TimothyEDawson committed:
Update advance_chemistry_FPV for new FluidPhysics.

TimothyEDawson committed:
Fix sound_speed call in plotting.

TimothyEDawson committed:
Forgot to remove n_scalar attribute in CanteraInterface.
